### PR TITLE
Fix zlib (de)compressor memory leaks

### PR DESCRIPTION
**Pull request by savilli**
`deflateEnd/inflateEnd` were not called on serialization errors. Let's use RAII and call them in destructors. 

## To do

This PR is Ready for Review.

### CODE REVIEW

File: src/serialization.cpp [hunk 1]
```diff
@@ -52,83 +52,119 @@ void zerr(int ret)
...
 
+class CompressZStream {
+public:
+	CompressZStream(int level) {
+		z.zalloc = Z_NULL;
+		z.zfree = Z_NULL;
+		z.opaque = Z_NULL;
+
+		int ret = deflateInit(&z, level);
+		if(ret != Z_OK)
```
Comment (Desour) [inline src/serialization.cpp:63]:
```suggestion
		if (ret != Z_OK)
```

File: src/serialization.cpp [hunk 2]
```diff
@@ -52,83 +52,119 @@ void zerr(int ret)
...
+			throw SerializationError("compressZlib: deflateInit failed");
+	}
+
+	~CompressZStream() {
+		deflateEnd(&z);
+	}
+
+	z_stream* get() {
```
Comment (Desour) [inline src/serialization.cpp:71]:
```suggestion
	z_stream *get() {
```

File: src/serialization.cpp [hunk 3]
```diff
@@ -52,83 +52,119 @@ void zerr(int ret)
     }
 }
 
+class CompressZStream {
```
Comment (Desour) [inline src/serialization.cpp:55]: Could use `DISABLE_CLASS_COPY()` from `util/basic_macros.h`.

**Review by Desour — COMMENTED** [review]
- :+1: For using RAII here.
- See https://dev.minetest.net/Code_style_guidelines for code style (sadly the wiki is still kinda broken).
- Code duplication of the two new classes could be reduced via templates, but I wouldn't suggest it due to simplicity.
  You could move the two classes next to each other to make it harder to not see that there's nearly identical code when making a change.
- Otherwise fine.

**Comment by sfan5** [comment]
Fixing this without involving C++ is 3 or 4 lines so I'm not quite sure if we really should.

**Comment by appgurueu** [comment]
> Fixing this without involving C++ is 3 or 4 lines so I'm not quite sure if we really should.

Please fix it. Should I open an issue?

**Comment by savilli** [comment]
Implemented another approach that requires fewer code changes

**Comment by appgurueu** [comment]
This is particularly important if data received over the wire (HTTP API) is decompressed to prevent possible DoS if malicious or simply corrupted compressed data is received.

**Comment by sfan5** [comment]
Let's code golf this some more then:
```diff
diff --git a/src/serialization.cpp b/src/serialization.cpp
index 11164a0ed..efcf09d3e 100644
--- a/src/serialization.cpp
+++ b/src/serialization.cpp
@@ -25,7 +25,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include <zstd.h>
 
 /* report a zlib or i/o error */
-void zerr(int ret)
+static void zerr(int ret)
 {
     dstream<<"zerr: ";
     switch (ret) {
@@ -52,6 +52,14 @@ void zerr(int ret)
     }
 }
 
+template <typename T, int (*D)(T*)>
+class AutoDeleter {
+	T *ptr;
+public:
+	AutoDeleter(T *ptr) : ptr(ptr) {}
+	~AutoDeleter() { D(ptr); }
+};
+
 void compressZlib(const u8 *data, size_t data_size, std::ostream &os, int level)
 {
 	z_stream z;
@@ -68,6 +76,8 @@ void compressZlib(const u8 *data, size_t data_size, std::ostream &os, int level)
 	if(ret != Z_OK)
 		throw SerializationError("compressZlib: deflateInit failed");
 
+	AutoDeleter<z_stream, deflateEnd> deleter(&z);
+
 	// Point zlib to our input buffer
 	z.next_in = (Bytef*)&data[0];
 	z.avail_in = data_size;
@@ -91,8 +101,6 @@ void compressZlib(const u8 *data, size_t data_size, std::ostream &os, int level)
 		if(status == Z_STREAM_END)
 			break;
 	}
-
-	deflateEnd(&z);
 }
 
 void compressZlib(const std::string &data, std::ostream &os, int level)
@@ -119,6 +127,8 @@ void decompressZlib(std::istream &is, std::ostream &os, size_t limit)
 	if(ret != Z_OK)
 		throw SerializationError("dcompressZlib: inflateInit failed");
 
+	AutoDeleter<z_stream, inflateEnd> deleter(&z);
+
 	z.avail_in = 0;
 
 	for(;;)
@@ -180,8 +190,6 @@ void decompressZlib(std::istream &is, std::ostream &os, size_t limit)
 			break;
 		}
 	}
-
-	inflateEnd(&z);
 }
 
 struct ZSTD_Deleter {
```

**Comment by savilli** [comment]
> Let's code golf this some more then:

Done

File: src/serialization.cpp [hunk 4]
```diff
@@ -52,6 +52,17 @@ void zerr(int ret)
...
 
+// Make sure that z is deleted in case of exception
+template <int (*F)(z_stream*)>
+class ZlibAutoDeleter {
+public:
+	ZlibAutoDeleter(z_stream *ptr) : ptr_(ptr) {}
+	~ZlibAutoDeleter() { F(ptr_); }
+
+private:
+	z_stream *ptr_;
```
Comment (sfan5) [inline src/serialization.cpp:63]: (this is equivalent)
```suggestion
class ZlibAutoDeleter {
	z_stream *ptr;
public:
	ZlibAutoDeleter(z_stream *ptr) : ptr(ptr) {}
	~ZlibAutoDeleter() { F(ptr); }
```

**Review by sfan5 — APPROVED** [review]

File: src/serialization.cpp (hunk 4 above)
Comment (savilli) [reply src/serialization.cpp]: This is stupid an unnecessary. Why shouldn't I change it back?

Comment (sfan5) [reply src/serialization.cpp]: This is both one line shorter and clearer (no unnecessary `ptr` `ptr_` distinction)

Comment (savilli) [reply src/serialization.cpp]: I don't consider one line shorter as a valid point. Follow code style and leave your sense of beauty for yourself. Thanks.

Comment (sfan5) [reply src/serialization.cpp:63]: Here are our code style guidelines, I don't see it mentioned: https://dev.minetest.net/Code_style_guidelines

> I don't consider one line shorter as a valid point.

What about the `ptr`, `ptr_` naming then?

Comment (savilli) [reply src/serialization.cpp:63]: It was mentioned in the beginning. Minetest style guide doesn't say anything about private members naming, but google style guide says they should have trailing underscoreshttps://google.github.io/styleguide/cppguide.html#Variable_Names

Comment (sfan5) [reply src/serialization.cpp:63]: Well we don't use use Google's style guide. Searching through the source for `_;` I find that jsoncpp does this consistently and most other uses are in constructors but the other way around (as in `C::C(T *foo_) { foo = foo_; }`).
So I don't think new code should use a style that does not fit existing code.

Comment (Zughy) [reply src/serialization.cpp:63]: @savilli

**Review by SmallJoker — APPROVED** [review]
I don't care about this trivial variable naming. LGTM.

**Pull request merged**